File: src/common/src/secrets/file-watcher.ts
import * as fs from 'node:fs';
import * as path from 'node:path';

export type RenameHandler = (filename: string) => void;

/**
 * Watches a directory tree for rename events by placing a non-recursive
 * `fs.watch` on every real (non-symlink) subdirectory.
 *
 * Why not `{ recursive: true }`?
 * Node.js `fs.watch({ recursive: true })` on Linux uses inotify but its
 * delivery of `IN_MOVED_TO` events for files inside *subdirectories* is
 * unreliable on container filesystems (e.g. overlayfs / Docker). Plain
 * per-directory inotify watches are fully supported on all Linux kernels and
 * filesystem types, and work identically on macOS (kqueue EVFILT_VNODE).
 *
 * Layout covered:
 *  - Root directory: detects `..data` symlink-swap (CSI driver rotation).
 *  - Each real subdirectory (e.g. `db/`, `tls/server/`): detects individual
 *    key-file atomic renames.
 *  - New subdirectories created at runtime are automatically picked up.
 */
export class DirectoryWatcher {
  private readonly watchers = new Map<string, fs.FSWatcher>();
  private closed = false;

  constructor(
    private readonly dir: string,
    private readonly onRename: RenameHandler,
    private readonly onWatchError: (err: Error) => void = (err) =>
      console.error('[secrets-client] fs.watch error:', err),
  ) {
    this._watch(dir, '');
  }

  private _watch(dirPath: string, prefix: string): void {
    if (this.watchers.has(dirPath)) return;

    const watcher = fs.watch(
      dirPath,
      { persistent: false },
      (eventType, filename) => {
        if (this.closed || eventType !== 'rename') return;

        if (!filename) {
          // Some platforms/filesystems omit the filename in the callback.
          // Conservatively re-scan the directory: signal the ..data symlink
          // for the root (covers CSI rotation) and every visible non-".."
          // entry so any pending rotation is picked up.
          if (!prefix) {
            this.onRename(path.join(this.dir, '..data'));
          }
          try {
            for (const entry of fs.readdirSync(dirPath)) {
              if (entry.startsWith('..')) continue;
              const relative = prefix ? `${prefix}/${entry}` : entry;
              this.onRename(path.join(this.dir, relative));
              try {
                if (fs.lstatSync(path.join(dirPath, entry)).isDirectory()) {
                  this._watch(path.join(dirPath, entry), relative);
                }
              } catch {
                // Entry disappeared — skip.
              }
            }
          } catch {
            // Directory gone or not listable — ignore.
          }
          return;
        }

        const relative = prefix ? `${prefix}/${filename}` : filename;
        this.onRename(path.join(this.dir, relative));

        // If a new real subdirectory just appeared, start watching it —
        // unless its name starts with ".." (CSI versioned data dirs such as
        // "..data_v2" or "..2026_01_01_..."). Those are internal driver
        // artefacts: we detect their rotation via the root-level "..data"
        // symlink rename and do not need to recurse into them. Skipping them
        // prevents inotify handle exhaustion as rotations accumulate over time.
        try {
          if (
            !filename.startsWith('..') &&
            fs.lstatSync(path.join(dirPath, filename)).isDirectory()
          ) {
            this._watch(path.join(dirPath, filename), relative);
          }
        } catch {
          // Entry is gone (rename source) or inaccessible — ignore.
        }
      },
    );

    watcher.on('error', (err) => {
      if (!this.closed) this.onWatchError(err as Error);
    });

    this.watchers.set(dirPath, watcher);

    // Recursively watch real subdirectories that already exist, skipping ".."
    // directories for the same reason as above.
    try {
      for (const entry of fs.readdirSync(dirPath)) {
        try {
          if (entry.startsWith('..')) continue;
          const entryPath = path.join(dirPath, entry);
          if (fs.lstatSync(entryPath).isDirectory()) {
            this._watch(entryPath, prefix ? `${prefix}/${entry}` : entry);
          }
        } catch {
          // Entry might be inaccessible — skip.
        }
      }
    } catch {
      // dirPath might not be listable — ignore.
    }
  }

  close(): void {
    this.closed = true;
    for (const w of this.watchers.values()) w.close();
    this.watchers.clear();
  }

  get directory(): string {
    return this.dir;
  }
}

/**
 * A debounced rename handler that also exposes `cancel()` so that callers
 * (e.g. `SecretStoreImpl.close()`) can drain any pending timer on shutdown,
 * honouring the "releases all timers" contract in `types.ts`.
 */
export interface DebouncedRenameHandler {
  (filePath: string): void;
  cancel(): void;
}

/**
 * Returns a debounced wrapper around `fn` that accumulates every unique path
 * seen within `delayMs` and then calls `fn` once for each unique path.
 *
 * Unlike a last-arg-wins debounce, this guarantees that the DESTINATION of an
 * atomic rename (e.g. `db/url`) is always processed even when the platform
 * fires the source event (`db/.tmp-xxx`) last. On macOS (FSEvents) and on
 * Linux with inotify, a `renameSync(src, dst)` produces two 'rename' events
 * whose delivery order is non-deterministic; accumulating both ensures the
 * cached key path reaches `_handleRename` regardless of that order.
 *
 * Call `.cancel()` to clear any pending timer and discard buffered paths.
 */
export function debounce(
  fn: (filePath: string) => void,
  delayMs: number,
): DebouncedRenameHandler {
  let timer: ReturnType<typeof setTimeout> | undefined;
  const pending = new Set<string>();

  const handler = (filePath: string) => {
    pending.add(filePath);
    clearTimeout(timer);
    timer = setTimeout(() => {
      const paths = [...pending];
      pending.clear();
      for (const p of paths) fn(p);
    }, delayMs);
  };

  handler.cancel = () => {
    clearTimeout(timer);
    pending.clear();
    timer = undefined;
  };

  return handler as DebouncedRenameHandler;
}

File: src/common-go/secrets-client/file_watcher.go
package secretsclient

import (
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/fsnotify/fsnotify"
)

// renameHandler is called with the absolute path of a file involved in a
// rename/create event.
type renameHandler func(filePath string)

// directoryWatcher places a non-recursive fsnotify watch on the root directory
// and every real (non-symlink) subdirectory, mirroring the TypeScript and
// Python DirectoryWatcher implementations.
//
// Directories whose names start with ".." (CSI driver versioned data
// directories such as "..data_v2") are deliberately skipped to avoid inotify
// handle exhaustion on long-lived clusters. The "..data" symlink-swap
// rotation is detected at the root level instead.
type directoryWatcher struct {
	root     string
	watcher  *fsnotify.Watcher
	onRename renameHandler
	onError  func(err error)
	watched  map[string]struct{}
	mu       sync.Mutex
	closed   bool
}

func newDirectoryWatcher(
	root string,
	onRename renameHandler,
	onError func(err error),
) (*directoryWatcher, error) {
	w, err := fsnotify.NewWatcher()
	if err != nil {
		return nil, err
	}
	dw := &directoryWatcher{
		root:     root,
		watcher:  w,
		onRename: onRename,
		onError:  onError,
		watched:  make(map[string]struct{}),
	}
	if err := dw.watchDir(root); err != nil {
		_ = w.Close()
		return nil, err
	}
	go dw.loop()
	return dw, nil
}

func (dw *directoryWatcher) watchDir(dir string) error {
	dw.mu.Lock()
	if _, ok := dw.watched[dir]; ok {
		dw.mu.Unlock()
		return nil
	}
	dw.watched[dir] = struct{}{}
	dw.mu.Unlock()

	if err := dw.watcher.Add(dir); err != nil {
		return err
	}

	entries, _ := os.ReadDir(dir)
	for _, entry := range entries {
		if strings.HasPrefix(entry.Name(), "..") {
			continue
		}
		// Use Lstat (not entry.Info which follows symlinks) so symlinked
		// directories are detected and excluded correctly.
		info, err := os.Lstat(filepath.Join(dir, entry.Name()))
		if err != nil {
			continue
		}
		// Only recurse into real (non-symlink) directories.
		if info.IsDir() && info.Mode()&os.ModeSymlink == 0 {
			_ = dw.watchDir(filepath.Join(dir, entry.Name()))
		}
	}
	return nil
}

func (dw *directoryWatcher) loop() {
	for {
		select {
		case event, ok := <-dw.watcher.Events:
			if !ok {
				return
			}
			dw.mu.Lock()
			closed := dw.closed
			dw.mu.Unlock()
			if closed {
				return
			}
			// Handle RENAME (source path) and CREATE (destination path).
			// Together these cover the full atomic-rename event pair on Linux
			// (inotify IN_MOVED_FROM + IN_MOVED_TO).
			if event.Has(fsnotify.Rename) || event.Has(fsnotify.Create) {
				// If a real non-".." subdirectory just appeared, start watching it.
				if info, err := os.Lstat(event.Name); err == nil &&
					info.IsDir() && info.Mode()&os.ModeSymlink == 0 &&
					!strings.HasPrefix(filepath.Base(event.Name), "..") {
					_ = dw.watchDir(event.Name)
				}
				dw.onRename(event.Name)
			}

		case err, ok := <-dw.watcher.Errors:
			if !ok {
				return
			}
			dw.mu.Lock()
			closed := dw.closed
			dw.mu.Unlock()
			if !closed && dw.onError != nil {
				dw.onError(err)
			}
		}
	}
}

func (dw *directoryWatcher) close() {
	dw.mu.Lock()
	dw.closed = true
	dw.mu.Unlock()
	_ = dw.watcher.Close()
}

// debouncedHandler accumulates every unique file path seen within delay and
// then calls fn once for each. Unlike last-arg-wins debounce, this guarantees
// that both the source and destination of an atomic rename are processed
// regardless of delivery order.
type debouncedHandler struct {
	fn      func(string)
	delay   time.Duration
	mu      sync.Mutex
	pending map[string]struct{}
	timer   *time.Timer
}

func newDebouncedHandler(fn func(string), delayMs int) *debouncedHandler {
	return &debouncedHandler{
		fn:      fn,
		delay:   time.Duration(delayMs) * time.Millisecond,
		pending: make(map[string]struct{}),
	}
}

func (d *debouncedHandler) call(filePath string) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.pending[filePath] = struct{}{}
	if d.timer != nil {
		d.timer.Stop()
	}
	d.timer = time.AfterFunc(d.delay, d.flush)
}

func (d *debouncedHandler) flush() {
	d.mu.Lock()
	paths := make([]string, 0, len(d.pending))
	for p := range d.pending {
		paths = append(paths, p)
	}
	d.pending = make(map[string]struct{})
	d.timer = nil
	d.mu.Unlock()
	for _, p := range paths {
		d.fn(p)
	}
}

func (d *debouncedHandler) cancel() {
	d.mu.Lock()
	defer d.mu.Unlock()
	if d.timer != nil {
		d.timer.Stop()
		d.timer = nil
	}
	d.pending = make(map[string]struct{})
}
